### PR TITLE
Rework proxy creation

### PR DESCRIPTION
**Pull request by pepone**

### CODE REVIEW

File: cpp/test/Ice/ami/AllTests.cpp [hunk 1]
```diff
@@ -80,17 +80,8 @@ allTests(Test::TestHelper* helper, bool collocated)
...
-
-    auto p = Ice::uncheckedCast<Test::TestIntfPrx>(prx);
-
-    sref = "testController:" + helper->getTestEndpoint(1);
-    prx = communicator->stringToProxy(sref);
-    test(prx);
-
-    auto testController = Ice::uncheckedCast<Test::TestIntfControllerPrx>(prx);
+    Test::TestIntfPrx p(communicator, "test:" + helper->getTestEndpoint());
+    Test::TestIntfControllerPrx testController(communicator, "testController:" + helper->getTestEndpoint(1));
```
Comment (pepone) [inline cpp/test/Ice/ami/AllTests.cpp:85]: This is the preferred way to create the proxy with the new mapping

File: cpp/test/Ice/ami/AllTests.cpp [hunk 2]
```diff
@@ -578,7 +569,7 @@ allTests(Test::TestHelper* helper, bool collocated)
 
     cout << "testing local exceptions with lambda API... " << flush;
     {
-        auto indirect = Ice::uncheckedCast<Test::TestIntfPrx>(p->ice_adapterId("dummy"));
+        Test::TestIntfPrx indirect = p->ice_adapterId("dummy");
```
Comment (pepone) [inline cpp/test/Ice/ami/AllTests.cpp:572]: This uncheckedCast is redundant.

File: cpp/test/Ice/ami/AllTests.cpp [hunk 3]
```diff
@@ -1099,7 +1088,7 @@ allTests(Test::TestHelper* helper, bool collocated)
             {
                 test(p->opBatchCount() == 0);
                 auto b1 = Ice::uncheckedCast<Test::TestIntfPrx>(
-                    p->ice_getConnection()->createProxy(p->ice_getIdentity())->ice_batchOneway());
+                    p->ice_getConnection()->createProxy(p->ice_getIdentity()))->ice_batchOneway();
```
Comment (pepone) [inline cpp/test/Ice/ami/AllTests.cpp:1091]: is there a better way to create a fixed proxy?

Comment (externl) [reply cpp/test/Ice/ami/AllTests.cpp:1091]: Use `ice_fixed`?

**Review by externl — APPROVED** [review]

File: cpp/test/Ice/ami/AllTests.cpp (hunk 3 above)
Comment (bernardnormier) [reply cpp/test/Ice/ami/AllTests.cpp:1091]: There is no reason to use uncheckedCast for anything. If you have an ObjectPrx proxy, you can create a typed proxy with the "const ObjectPrx&" ctor, as in:
```cpp
Test::TestInftPrx b1(p->ice_getConnection()->createProxy(p->ice_getIdentity()));
```

Comment (bernardnormier) [reply cpp/test/Ice/ami/AllTests.cpp:1091]: Here, I would indeed use ice_fixed:
```cpp
auto b1 = p->ice_fixed(p->ice_getConnection())->ice_batchOneway();
```

File: cpp/test/Ice/acm/TestI.h [hunk 4]
```diff
@@ -9,40 +9,40 @@
...
     RemoteObjectAdapterI(const Ice::ObjectAdapterPtr&);
 
-    virtual Test::TestIntfPrxPtr getTestIntf(const Ice::Current&);
-    virtual void activate(const Ice::Current&);
-    virtual void hold(const Ice::Current&);
-    virtual void deactivate(const Ice::Current&);
+    std::optional<Test::TestIntfPrx> getTestIntf(const Ice::Current&) final;
+    void activate(const Ice::Current&) final;
+    void hold(const Ice::Current&) final;
+    void deactivate(const Ice::Current&) final;
```
Comment (bernardnormier) [inline cpp/test/Ice/acm/TestI.h:29]: replace Test::TestIntfPrxPtr below by const Test::TestIntfPrx

File: cpp/test/Ice/admin/TestI.h [hunk 5]
```diff
@@ -9,28 +9,28 @@
 #include <TestHelper.h>
 #include <Ice/NativePropertiesAdmin.h>
 
-class RemoteCommunicatorI : public virtual Test::RemoteCommunicator
+class RemoteCommunicatorI final : public virtual Test::RemoteCommunicator
```
Comment (bernardnormier) [inline cpp/test/Ice/admin/TestI.h:12]: virtual is probably unecessary

**Review by bernardnormier — APPROVED** [review]

**Pull request merged**